bottlecap/src/traces/trace_agent.rs

```diff
@@ -148,50 +150,78 @@ impl TraceAgent {
-                    trace_tx.clone(),
-                    stats_processor.clone(),
-                    stats_tx.clone(),
-                    invocation_processor.clone(),
-                    tags_provider.clone(),
-                    client.clone(),
-                    api_key.clone(),
-                )
-            });
-
-            async move { Ok::<_, Infallible>(service) }
+            TraceAgent::trace_endpoint_handler(
+                endpoint_config.clone(),
+                req.map(hyper_migration::Body::incoming),
+                trace_processor.clone(),
+                trace_tx.clone(),
+                stats_processor.clone(),
+                stats_tx.clone(),
+                invocation_processor.clone(),
+                tags_provider.clone(),
+                client.clone(),
+                api_key.clone(),
+            )
         });
 
         let port = u16::try_from(TRACE_AGENT_PORT).expect("TRACE_AGENT_PORT is too large");
         let addr = SocketAddr::from(([127, 0, 0, 1], port));
-        let server_builder = Server::try_bind(&addr)?;
-
-        let server = server_builder.serve(make_svc);
+        let listener = tokio::net::TcpListener::bind(&addr).await?;
 
         debug!("Trace Agent started: listening on port {TRACE_AGENT_PORT}");
         debug!(
-            "Time taken start the Trace Agent: {} ms",
+            "Time taken start the Mini Agent: {} ms",
             now.elapsed().as_millis()
         );
-
-        // start hyper http server
-        if let Err(e) = server.await {
-            error!("Server error: {e}");
-            return Err(e.into());
+        let server = hyper::server::conn::http1::Builder::new();
+        let mut joinset = tokio::task::JoinSet::new();
+        loop {
+            let conn = tokio::select! {
+                con_res = listener.accept() => match con_res {
+                    Err(e)
+                        if matches!(
+                            e.kind(),
+                            io::ErrorKind::ConnectionAborted
+                                | io::ErrorKind::ConnectionReset
+                                | io::ErrorKind::ConnectionRefused
+                        ) =>
+                    {
+                        continue;
+                    }
+                    Err(e) => {
+                        error!("Server error: {e}");
+                        return Err(e.into());
+                    }
+                    Ok((conn, _)) => conn,
+                },
+                finished = async {
+                    match joinset.join_next().await {
+                        Some(finished) => finished,
+                        None => std::future::pending().await,
+                    }
+                } => match finished {
+                    Err(e) if e.is_panic() => {
+                        std::panic::resume_unwind(e.into_panic());
+                    },
+                    Ok(()) | Err(_) => continue,
+                },
+            };
```

Comment:
Looks like we have the same thing in other places, does it make sense to make it a method/helper?

Also took me a while to understand this, too many matches, should we add some comments?

Reply:
It's just a server loop. I can refactor it in another PR